string data
int32 counter
int32[] error
